### PR TITLE
ship_protocol

### PR DESCRIPTION
**Pull request by tbfleming**

### CODE REVIEW

File: include/eosio/ship_protocol.hpp [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+         return transaction_status::delayed;
+      if (s == "expired")
+         return transaction_status::expired;
+      report_error("unknown status: " + s);
+   }
+
+   template <typename S>
+   eosio::result<void> from_bin(transaction_status& obj, S& stream) {
+      return stream.read_raw(obj);
+   }
```
Comment (swatanabe-b1) [inline include/eosio/ship_protocol.hpp]: I don't think this is needed.  This is the default behavior for scoped enums.

File: include/eosio/ship_protocol.hpp [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+      eosio::name                        receiver               = {};
+      action                             act                    = {};
+      bool                               context_free           = {};
+      int64_t                            elapsed                = {};
+      std::string                        console                = {};
+      std::vector<account_delta>         account_ram_deltas     = {};
+      std::vector<account_delta>         account_disk_deltas    = {};
+      std::optional<std::string>         except                 = {};
+      std::optional<uint64_t>            error_code             = {};
+      std::optional<eosio::input_stream> return_value           = {};
```
Comment (swatanabe-b1) [inline include/eosio/ship_protocol.hpp]: The final cut of the action return value in state_history_plugin doesn't use optional. https://github.com/EOSIO/eos/blob/2db44d405629c0e4182e013cfff1d6f0bbe4b130/plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_serialization.hpp#L598

File: include/eosio/ship_protocol.hpp (hunk 1 above)
Comment (tbfleming) [reply include/eosio/ship_protocol.hpp]: removed

File: include/eosio/ship_protocol.hpp (hunk 2 above)
Comment (tbfleming) [reply include/eosio/ship_protocol.hpp]: changed

**Pull request merged**